server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -149,10 +150,29 @@ protected WritePrimaryResult<ShardUpsertRequest, ShardResponse> processRequestIt
         TransactionContext txnCtx = TransactionContext.of(request.sessionSettings());
         UpdateToInsert updateToInsert = request.updateColumns() == null || request.updateColumns().length == 0
             ? null
-            : new UpdateToInsert(nodeCtx, txnCtx, tableInfo, request.updateColumns());
+            : new UpdateToInsert(
+                nodeCtx,
+                txnCtx,
+                tableInfo,
+                request.updateColumns(),
+                request.insertColumns()
+            );
         List<Reference> insertColumns;
         if (updateToInsert != null) {
             insertColumns = updateToInsert.columns();
+
+            // updateToInsert.columns() must have an exact overlap with the insert
+            // target-column-list because the supplied values
+            // will be in that order.
+            //
+            // Example where it adds columns:
+            //
+            // INSERT INTO tbl (x) VALUES (1)
```

Comment:
Just a suggestion: maybe it's better to use an example with:
```
INSERT INTO tbl (y, z) VALUES (1, 2)
   ON CONFLICT (y) DO UPDATE SET x = 20
```
So that below we can show: `[z, y, ...]`

Reply:
The example is for adding columns, not to illustrate the column swap behavior, and that problem could also happen with a single column.